specs/plugin-extensibility.md

```diff
@@ -0,0 +1,277 @@
+* Every request JSON will contain a `contract-version` top level attribute whose value will indicate the plugin contract version. Contract version is revised when there are changes to command request/response, new plugin commands are introduced, and supported through Notation.
+* For an error response, every command returns a non-zero exit code 1, with a standard JSON error response in `stderr`. There is no need to send different exit codes for different error conditions, as Notation (which will call the plugin and parse the response) will use `error-code` in the error response to intepret different error conditions if needed.
+
+```json
+{
+  // Each command defines expected error codes
+  "error-code" : "<error code>",  
+  // Plugin defined error message.
+  "error-message" : "User friendly error message"
+
+  // Optional plugin defined additional key value pairs related to the error.
+  "error-metadata" : {
+    "key1" : "value1",
+    "key2" : "value2"
+  }
+}
+```
+
+### Plugin metadata
+
+* Every plugin MUST implement a metadata dicovery command called `discover`. 
+* Notation will invoke this command as part of signing and verification workflows to discover the capabilities of the plugin and subsequently invoke commands corresponding to the capabilities. Notation will invoke the `discover` command every time a signing or verification workflow is invoked to discover metadata. Notation will not cache the response of `discover` command as it involves invalidating cache when a plugin is updated, detecting a plugin update can be non trivial. The `discover` command is expected to return only static data, and plugin implementors should avoid making remote calls in this command.
+
+***discover***
+
+*Request* - None
+
+*Response*
+All response attributes are required.
+
+```json
+{
+  // Plugin name that matches its install dir
+  // e.g. "com.example.nv2plugin".
+  "plugin-name" : "<plugin name>",  
+  // Plugin friendly name.
+  "plugin-friendly-name" : "<friendly name>",
+  // Plugin publisher controlled version.
+  "plugin-version" : "<version>",
+  // Plugin webpage for support or documentation.
+  "plugin-url" : "<URL>",
+  
+  // Contract versions supported by the plugin
+  "supported-contract-versions" : "<contract-version>",
+  
+  // Currently one of 
+  // SIGNATURE_GENERATOR or
+  // SIGNATURE_ENVELOPE_GENERATOR
```

Comment:
If `SIGNATURE_GENERATOR` and `SIGNATURE_ENVELOPE_GENERATOR` are supported by the plugin which one takes precedence when `notation sign` is invoked?

Reply:
Currently we expect a signing plugin to implement one of the signing interfaces not both. Let me know if this does not work for a specific case.